### PR TITLE
GPT-3 Command for Pwndbg

### PR DESCRIPTION
**Pull request by oblivia-simplex**
I've added the `ai` command that will let the user ask OpenAI's GPT-3 LLM questions about the current debugging context or the output of a particular command, and sustain a brief conversation with the model on the topic by asking follow-up questions. The environment variable `OPENAI_API_KEY` must contain a valid OpenAI API key for this command to work.

### CODE REVIEW

File: pwndbg/commands/ai.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
+"""
+This command sends information on the current debugging context to OpenAI's
+GPT-3 large language model and asks it a question supplied by the user. It then
+displays GPT-3's response to that question to the user.
+"""
+
+__AUTHOR__ = "Olivia Lucca Fraser"
+__VERSION__ = 0.1
+__LICENSE__ = "MIT"
```
Comment (gsingh93) [inline pwndbg/commands/ai.py]: @disconnect3d thoughts on including this info in the file?

File: pwndbg/commands/ai.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+import gdb
+import openai
+
+import pwndbg
+import pwndbg.commands
+from pwndbg.commands import CommandCategory
+from pwndbg.commands import context
+from pwndbg.gdblib import regs as REGS
+
+OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
```
Comment (gsingh93) [inline pwndbg/commands/ai.py]: Maybe we have a `Config` option for this that defaulted to the environment variable instead? Then I can just keep it in my .gdbinit.

File: pwndbg/commands/ai.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+LAST_QUESTION = []
+LAST_ANSWER = []
+HISTORY_LENGTH = 3
+LAST_PC = None
+STACK_DEPTH = 16
+LAST_COMMAND = None
```
Comment (gsingh93) [inline pwndbg/commands/ai.py]: nit: our coding style is lowercase snake, i.e. `last_question = []`

File: pwndbg/commands/ai.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+def build_prompt(question, command=None):
+    if command is not None:
+        return build_prompt_from_command(question, command)
+
+    decompile = False
+    ## First, get the current GDB context
+    ## Let's begin with the assembly near the current instruction
+    try:
+        # asm = gdb.execute("capstone-disassemble --length 32 $pc", to_string=True)
```
Comment (gsingh93) [inline pwndbg/commands/ai.py]: I guess we can remove this line?

File: pwndbg/commands/ai.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        asm_rows = pwndbg.gdblib.nearpc.nearpc(emulate=True, lines=16)
+        asm = "\n".join(asm_rows)
+    except Exception as e:
+        print(f"Error: {e}")
```
Comment (gsingh93) [inline pwndbg/commands/ai.py]: We can add some color to these error messages:
```
import pwndbg.color.message as M

print(M.error(f"Error: {e}"))
```

File: pwndbg/commands/ai.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    except Exception:
+        pass
+    if flags:
+        # just grab what's bewteen the square brackets
+        try:
+            flags = re.search(r"\[(.*)\]", flags).group(1)
+        except Exception:
+            pass
+    ## Finally, let's get the stack
+    # stack = gdb.execute("x/16x $sp", to_string=True)
```
Comment (gsingh93) [inline pwndbg/commands/ai.py]: We can remove this line

File: requirements.txt [hunk 7]
```diff
@@ -10,3 +10,5 @@ typing-extensions==4.3.0; python_version >= '3.7'
 typing-extensions==4.1.1; python_version < '3.7'
 unicorn==2.0.1.post1; python_version >= '3.7'
 unicorn==2.0.0rc7; python_version < '3.7'
+openai>=0.10.0
```
Comment (gsingh93) [inline requirements.txt]: I'm thinking this shouldn't be a required dependency. @disconnect3d ?

**Review by gsingh93 — COMMENTED** [review]
This is a very cool idea, thanks for implementing it.

If it's going to be too much of a hassle getting this to work on 18.04, we can just limit this to Ubuntu 20.04 and higher.

File: pwndbg/commands/ai.py (hunk 1 above)
Comment (disconnect3d) [reply pwndbg/commands/ai.py]: I'd prefer to not include it, if its fine with @oblivia-simplex

**Comment by oblivia-simplex** [comment]
Okay, it looks like the arch docker job is failing for unrelated reasons, so we're probably good to go?

File: requirements.txt [hunk 8]
```diff
@@ -10,3 +10,4 @@ typing-extensions==4.3.0; python_version >= '3.7'
 typing-extensions==4.1.1; python_version < '3.7'
 unicorn==2.0.1.post1; python_version >= '3.7'
 unicorn==2.0.0rc7; python_version < '3.7'
+requests>=2.0.0
```
Comment (disconnect3d) [inline requirements.txt]: Is there a reason why >=2.0.0 and not >=2.7.0 which is latest?

**Comment by disconnect3d** [comment]
Btw:
```
pwndbg/commands/ai.py:14: error: Library stubs not installed for "requests" 
[import]
    import requests
    ^
pwndbg/commands/ai.py:14: note: Hint: "python3 -m pip install types-requests"
pwndbg/commands/ai.py:14: note: (or run "mypy --install-types" to install all missing stub packages)
```

Shall we install `types-requests` in dev-requirements.txt?

File: pwndbg/commands/ai.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+    if True or asm:
+        prompt += f"""These are the next assembly instructions to be executed:
+
+```
+{asm}
+```
+
+"""
+    if True or regs:
```
Comment (gsingh93) [inline pwndbg/commands/ai.py]: What's with the `if True`?

**Comment by CptGibbon** [comment]
I'd love to see an example of this in action, perhaps just a screenshot?

**Comment by disconnect3d** [comment]
> I'd love to see an example of this in action, perhaps just a screenshot?

It's there in examples; but we can probably have a better example. Maybe we can ask AI to explain the code that checks the stack canary and branches off to _stack_chk_fail?

**Comment by oblivia-simplex** [comment]
Sure thing, I'm actually in the process of putting together a blog post on
this tool for work, and I'll get a handful of interesting examples prepared
for that. Once I do, I'll copy them into or link to them from the docs.

**Pull request merged**